lib/tests/streamlit/runtime/metrics_util_test.py

```diff
@@ -211,3 +210,157 @@ def test_function(param1: int, param2: str, param3: float = 0.1) -> str:
+            f"Command {expected_name} was not tracked.",
+        )
+
+    def test_public_api_commands(self):
+        """All commands of the public API should be tracked with the correct name."""
+        ctx = get_script_run_ctx()
+        ctx.reset()
+        ctx.gather_usage_stats = True
+
+        # Some commands are currently not tracked for various reasons:
+        ignored_commands = [
+            "experimental_rerun",
+            "stop",
+            "spinner",
+            "empty",
+            "progress",
+            "get_option",
+        ]
+        public_commands = {
+            k
+            for k, v in st.__dict__.items()
+            if not k.startswith("_") and not isinstance(v, type(st))
+        }
+
+        for command_name in public_commands:
+            if command_name in ignored_commands:
+                continue
+            command = getattr(st, command_name)
+            if callable(command):
+                # This will always throw an exception because of missing arguments
+                # This is fine since the command still get tracked
+                with contextlib.suppress(Exception):
+                    command()
+
+                self.assertGreater(
+                    len(ctx.tracked_commands),
+                    0,
+                    f"No command tracked for {command_name}",
+                )
+
+                # Sometimes also multiple commands are executed
+                # so we check the full list.
+                self.assertIn(
+                    command_name,
+                    [
+                        tracked_commands.name
+                        for tracked_commands in ctx.tracked_commands
+                    ],
+                )
+
+                ctx.reset()
+                ctx.gather_usage_stats = True
+
+    def test_command_tracking_limits(self):
+        """Command tracking limits should be respected.
+
+        Current limits are 25 per unique command and 200 in total.
+        """
+        ctx = get_script_run_ctx()
+        ctx.reset()
+        ctx.gather_usage_stats = True
+
+        @metrics_util.gather_metrics
+        def test_function1() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function2() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function3() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function4() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function5() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function6() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function7() -> str:
+            return "foo"
+
+        @metrics_util.gather_metrics
+        def test_function8() -> str:
+            return "foo"
```

Comment:
Could we create these functions in a loop?

```python
funcs = []
for ii in range(8):
  def test_function() -> str:
    return "foo"
  funcs.append(metrics_util.gather_metrics(test_function, name=f"test_function_{ii}"))
```

Reply:
Ahh thanks, that is much better :)